pyo3-ffi/src/object.rs

```diff
@@ -409,26 +431,32 @@ extern "C" {
 
 // Reference counting macros.
 #[inline]
+#[cfg(not(any(py_sys_config = "Py_REF_DEBUG", Py_NOGIL)))]
 pub unsafe fn Py_INCREF(op: *mut PyObject) {
-    if cfg!(py_sys_config = "Py_REF_DEBUG") {
-        Py_IncRef(op)
-    } else {
-        (*op).ob_refcnt += 1
-    }
+    (*op).ob_refcnt += 1
+}
+
+#[inline]
+#[cfg(any(py_sys_config = "Py_REF_DEBUG", Py_NOGIL))]
+pub unsafe fn Py_INCREF(op: *mut PyObject) {
+    Py_IncRef(op)
```

Comment:
@colesbury - I understood that for the CPython stable API there was a discussion about making refcounting details internal to CPython but it was deemed infeasible due to performance regression.

Here this patch seems to do exactly that. Are you able to comment on the estimated performance impact for extensions by changing refcounting to be through an FFI call?

Reply:
I think I have heard something similar, but I can't find the original source and don't know for which extensions it was deemed too great a cost.

There didn't seem to be an noticeable performance impact for the cryptography extension, but I don't know enough to make any general estimates.